File: CoreData/CoreDataConcurrencyDemo/CoreDataConcurrencyDemo/EditorViewController.m
//
//  EditorViewController.m
//  CoreDataConcurrencyDemo
//
//  Created by Tim Ekl on 2015.01.18.
//  Copyright (c) 2015 Tim Ekl. All rights reserved.
//

#import "EditorViewController.h"

#import <CoreData/CoreData.h>

@interface EditorViewController () <UITextFieldDelegate>

@property (nonatomic, strong) NSManagedObject *editingObject;

@property (weak, nonatomic) IBOutlet UITextField *stringField;

@end

@implementation EditorViewController

- (void)setContext:(NSManagedObjectContext *)context;
{
    _context = context;
    
    //add a new object to this context
    self.editingObject = [NSEntityDescription insertNewObjectForEntityForName:@"StringEntity" inManagedObjectContext:context];
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string;
{
    NSParameterAssert(textField == self.stringField);
    
    NSString *newStringValue = [[textField text] stringByReplacingCharactersInRange:range withString:string];
    
    //make changes to the object in the context
    [self.editingObject setValue:newStringValue forKey:@"string"];
    
    return YES;
}

@end

File: CoreData/CoreDataConcurrencyDemoSwift/CoreDataConcurrencyDemoSwift/EditorViewController.swift
//
//  EditorViewController.swift
//  CoreDataConcurrencyDemoSwift
//
//  Created by Ancil on 10/4/15.
//  Copyright © 2015 Ancil Marshall. All rights reserved.
//

import UIKit
import CoreData

class EditorViewController: UIViewController, UITextFieldDelegate {

    var object:NSManagedObject?
    var delegate:EditorViewControllerDelegate?
    @IBOutlet weak var inputText:UITextField!
    
    var context:NSManagedObjectContext? {
        didSet {
            object = NSEntityDescription.insertNewObjectForEntityForName("StringEntity", inManagedObjectContext: context!)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }
    
    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
        
            return true
    }
    
    
    @IBAction func save(){
        if let delegate = delegate {
            object?.setValue(inputText.text, forKey: "string")
            delegate.saveEditor()
            self.dismissViewControllerAnimated(true, completion: nil)
        }
    }
    
    @IBAction func cancel(){
        if let delegate = delegate {
            delegate.cancelEditor()
            self.dismissViewControllerAnimated(true, completion: nil)
        }
    }

}


protocol EditorViewControllerDelegate {
    
    func saveEditor()
    func cancelEditor()
}